plugins/synaptics-rmi/fu-synaptics-rmi-device.c

```diff
@@ -0,0 +1,1070 @@
+{
+	if (flash->bootloader_id[0] != 0x0) {
+		g_autofree gchar *tmp = g_strdup_printf ("%02x.%02x",
+							 flash->bootloader_id[0],
+							 flash->bootloader_id[1]);
+		fu_common_string_append_kv (str, idt, "BootloaderId", tmp);
+	}
+	fu_common_string_append_kx (str, idt, "BlockSize", flash->block_size);
+	fu_common_string_append_kx (str, idt, "BlockCountFw", flash->block_count_fw);
+	fu_common_string_append_kx (str, idt, "BlockCountCfg", flash->block_count_cfg);
+	fu_common_string_append_kx (str, idt, "FlashConfigLength", flash->config_length);
+	fu_common_string_append_kx (str, idt, "PayloadLength", flash->payload_length);
+	fu_common_string_append_kb (str, idt, "HasConfigID", flash->has_config_id);
+	fu_common_string_append_kb (str, idt, "HasNewRegmap", flash->has_new_regmap);
+	fu_common_string_append_kx (str, idt, "BuildID", flash->build_id);
+}
+
+static void
+fu_synaptics_rmi_device_to_string (FuDevice *device, guint idt, GString *str)
+{
+	FuSynapticsRmiDevice *self = FU_SYNAPTICS_RMI_DEVICE (device);
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	fu_common_string_append_ku (str, idt, "FD", (guint) priv->fd);
+	fu_common_string_append_kx (str, idt, "BlVer", priv->f34->function_version + 0x5);
+	fu_common_string_append_kx (str, idt, "ManufacturerID", priv->manufacturer_id);
+	fu_common_string_append_kb (str, idt, "HasLts", priv->has_lts);
+	fu_common_string_append_kb (str, idt, "HasSensorID", priv->has_sensor_id);
+	fu_common_string_append_kb (str, idt, "HasQuery42", priv->has_query42);
+	fu_common_string_append_kb (str, idt, "HasDS4Queries", priv->has_dds4_queries);
+	fu_common_string_append_kx (str, idt, "ConfigID", priv->config_id);
+	fu_common_string_append_kx (str, idt, "PackageID", priv->package_id);
+	fu_common_string_append_kx (str, idt, "PackageRev", priv->package_rev);
+	fu_common_string_append_kx (str, idt, "SensorID", priv->sensor_id);
+	fu_synaptics_rmi_flash_to_string (&priv->flash, idt, str);
+}
+
+FuSynapticsRmiFunction *
+fu_synaptics_rmi_device_get_function (FuSynapticsRmiDevice *self,
+				      guint8 function_number,
+				      GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	if (priv->functions->len == 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "no RMI functions, perhaps read the PDT?");
+		return NULL;
+	}
+	for (guint i = 0; i < priv->functions->len; i++) {
+		FuSynapticsRmiFunction *func = g_ptr_array_index (priv->functions, i);
+		if (func->function_number == function_number)
+			return func;
+	}
+	g_set_error (error,
+		     FWUPD_ERROR,
+		     FWUPD_ERROR_INTERNAL,
+		     "failed to get RMI function 0x%02x",
+		     function_number);
+	return NULL;
+}
+
+GByteArray *
+fu_synaptics_rmi_device_read (FuSynapticsRmiDevice *self, guint16 addr, gsize req_sz, GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	guint8 input_count_sz;
```

Comment:
not initializing makes me worry about compiler optimizations when it come to it's first use in the for loop.  Feel free to ignore if you think that's unfounded though.

Reply:
I think it's totally valid, and I'm pretty sure you'd hit that for the higher optimisation levels. Good catch!